src/server/pfs/server/url.go

```diff
@@ -100,70 +135,82 @@ func putFileURLRecursive(ctx context.Context, taskService task.Service, uw *file
 			if err != nil {
 				return err
 			}
-			return uw.AddFileSet(ctx, *fsid)
-		})
-		return errors.EnsureStack(err)
+			if err := uw.AddFileSet(ctx, *fsid); err != nil {
+				return errors.EnsureStack(err)
+			}
+		}
+		return nil
 	})
-	eg.Go(func() (retErr error) {
-		url, err := obj.ParseURL(src.URL)
-		if err != nil {
-			return errors.EnsureStack(err)
+	return errors.EnsureStack(eg.Wait())
+}
+
+// coordinateTasks iterates through a list of objects and creates tasks of uniform byte size by sharding/batching objects.
+// createTask is passed through as a callback to support nicer testing.
+func coordinateTasks(ctx context.Context, URL string, createTask func(startOffset, endOffset int64, startPath, endPath string) error) (retErr error) {
+	url, err := obj.ParseURL(URL)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	bucket, err := openBucket(ctx, url)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := bucket.Close(); err != nil {
+			retErr = multierror.Append(retErr, errors.EnsureStack(err))
 		}
-		bucket, err := openBucket(ctx, url)
-		if err != nil {
-			return err
+	}()
+	startPath, endPath := "", ""
+	list := bucket.List(&blob.ListOptions{Prefix: url.Object})
+	listObj, err := list.Next(ctx)
+	if err != nil {
```

Comment:
This doesn't seem right. A real error isn't returned and an EOF seems like shouldn't be an error.